lib/src/execution/windows/execution.cc

```diff
@@ -195,11 +195,151 @@ namespace facter { namespace execution {
+                    break;
+                }
+
+                // Read the pipe until pending
+                while (true) {
+                    // Before doing anything, check to see if there's been a timeout
+                    // This is done pre-emptively in case ReadFile never returns ERROR_IO_PENDING
+                    if (timeout && WaitForSingleObject(timer, 0) == WAIT_OBJECT_0) {
+                        throw timeout_exception((boost::format("command timed out after %1% seconds.") % timeout).str(), static_cast<size_t>(child));
+                    }
+
+                    // Read the data
+                    buffer->resize(4096);
+                    DWORD count = 0;
+                    if (!ReadFile(handle, &(*buffer)[0], buffer->size(), &count, overlapped)) {
+                        // Treat broken pipes as closed pipes
+                        if (GetLastError() == ERROR_BROKEN_PIPE) {
+                            handle = INVALID_HANDLE_VALUE;
+                            break;
+                        }
+                        // Check to see if it's a pending operation
+                        if (GetLastError() == ERROR_IO_PENDING) {
+                            pending = true;
+                            break;
+                        }
+                        LOG_ERROR("failed to read child output: %1%.", system_error());
+                        throw execution_exception("failed to read child process output.");
+                    }
+
+                    // Check for closed pipe
+                    if (count == 0) {
+                        handle = INVALID_HANDLE_VALUE;
+                        break;
+                    }
+
+                    // Read completed immediately, process the data
+                    buffer->resize(count);
+                    if (!(*callback)(*buffer)) {
+                        // Callback signaled that we're done
+                        return;
+                    }
+                }
+            }
+
+            // All pipes should be pending now
+            wait_handles.clear();
+            for (auto const& pipe : pipes) {
+                auto& handle = get<0>(pipe);
+                auto overlapped = get<1>(pipe);
+                auto& pending = get<2>(pipe);
+
+                if (handle == INVALID_HANDLE_VALUE || !pending) {
+                    continue;
+                }
+                wait_handles.push_back(overlapped->hEvent);
+            }
+
+            // If no wait handles, then we're done processing
+            if (wait_handles.empty()) {
+                return;
+            }
+
+            if (timeout) {
+                wait_handles.push_back(timer);
+            }
+
+            // Wait for data (and, optionally, timeout)
+            auto result = WaitForMultipleObjects(wait_handles.size(), wait_handles.data(), FALSE, INFINITE);
+            if (result >= (WAIT_OBJECT_0 + wait_handles.size())) {
```

Comment:
I was going to ask about assuming WAIT_OBJECT_0 is 0 and that result is unsigned, but the docs for WaitForMultipleObjects explicitly mention those things.